### PR TITLE
s3: avoid array copies when dealing with ByteStrings

### PR DESCRIPTION
**Pull request by pjfanning**
the 'unsafe' methods are ok if you know the data is not being mutated

### CODE REVIEW

**Comment by pjfanning** [comment]
@mdedetrich You are the most active contributor to the S3 code base. Do these changes look ok to you?

File: s3/src/main/scala/org/apache/pekko/stream/connectors/s3/impl/auth/package.scala [hunk 1]
```diff
@@ -36,7 +36,16 @@ package object auth {
     new String(out)
   }
 
-  @InternalApi private[impl] def encodeHex(bytes: ByteString): String = encodeHex(bytes.toArray)
+  @InternalApi private[impl] def encodeHex(bytes: ByteString): String = {
+    val length = bytes.length
```
Comment (pjfanning) [inline s3/src/main/scala/org/apache/pekko/stream/connectors/s3/impl/auth/package.scala:40]: ByteString implements `IndexedSeq[Byte]` so we can iterate it without converting it to an array.

Comment (pjfanning) [reply s3/src/main/scala/org/apache/pekko/stream/connectors/s3/impl/auth/package.scala:40]: looking at the ByteStrings code (complex ByteString made up of underlying ByteStrings) the apply(Int) function and the iterator function are not performant.

https://github.com/apache/pekko/blob/c4805f1839ead0cfe31700d6c41928df9fb1056d/actor/src/main/scala-3/org/apache/pekko/util/ByteString.scala#L554-L570

So I'm now thinking of changing this to
```
def encodeHex(bytes: ByteString): String = encodeHex(bytes.toArrayUnsafe)
```

**Review by nvollmar — APPROVED** [review]
lgtm

**Pull request merged**